crates/circuit/src/parameter_expression.rs

```diff
@@ -333,173 +1834,565 @@ impl ParameterExpression {
+        for (i, p) in params.iter().enumerate() {
+            if i != params.len() - 1 {
+                out += &format!("'{}', ", p);
+            } else {
+                out += &format!("'{}'", p);
+            }
+        }
+        out += "]";
+        out
+    }
+
+    pub fn __repr__(&self) -> String {
+        format!(
+            "ParameterVector(name='{}', length={})",
+            self.name,
+            self.params.len()
+        )
+        .to_string()
+    }
+
+    /// """Resize the parameter vector.  If necessary, new elements are generated.
+    ///
+    /// Note that the UUID of each :class:`.Parameter` element will be generated
+    /// deterministically given the root UUID of the ``ParameterVector`` and the index
+    /// of the element.  In particular, if a ``ParameterVector`` is resized to
+    /// be smaller and then later resized to be larger, the UUID of the later
+    /// generated element at a given index will be the same as the UUID of the
+    /// previous element at that index.
+    /// This is to ensure that the parameter instances do not change.
+    ///
+    /// >>> from qiskit.circuit import ParameterVector
+    /// >>> pv = ParameterVector("theta", 20)
+    /// >>> elt_19 = pv[19]
+    /// >>> rv.resize(10)
+    /// >>> rv.resize(20)
+    /// >>> pv[19] == elt_19
+    /// True
+    /// """
+    #[pyo3(name = "resize")]
+    pub fn py_resize(&mut self, length: usize) {
+        if length > self.params.len() {
+            let root_uuid = self.root_uuid;
+            for i in self.params.len()..length {
+                let pe = ParameterExpression {
+                    expr: SymbolExpr::Symbol {
+                        name: Box::new(self.name.clone()),
+                        index: Some(i),
+                    },
+                    uuid: root_uuid + i as u128,
+                    qpy_replay: None,
+                    parameter_symbols: None,
+                    parameter_vector: None,
+                };
+                self.params.push(pe);
+            }
+            let t = Arc::<ParameterVector>::new(self.to_owned());
+            for pe in &mut self.params {
+                pe.parameter_vector = Some(t.clone());
+            }
+        } else {
+            self.params.resize(length, ParameterExpression::default());
+        }
+    }
+
+    fn __getstate__(&self) -> PyResult<(String, usize, u128)> {
+        Ok((self.name.clone(), self.params.len(), self.root_uuid.clone()))
+    }
+    fn __setstate__(&mut self, state: (String, usize, u128)) -> PyResult<()> {
+        self.name = state.0;
+        self.root_uuid = state.2;
+        let length = state.1;
+        self.params = Vec::with_capacity(length);
+        for i in 0..length {
+            let pe = ParameterExpression {
+                expr: SymbolExpr::Symbol {
+                    name: Box::new(self.name.clone()),
+                    index: Some(i),
+                },
+                uuid: self.root_uuid + i as u128,
```

Comment:
This feels a bit odd. Presumably we do the same thing (some UUID + i) in Python?